Tests: add tests for non-root key rotations

MVrachev opened the pull request:
Fixes #1647

**Description of the changes being introduced by the pull request**:

The amount of non-root key rotations is lower than the one for root as
the repository stores all root versions in order to be able to do
rollback check compared to timestamp, snapshot, and targets where it
stores only one version.

All of the root roles are needed in those tests again as root
delegates the keys for each of the other three top-level metadata roles.

Additionally, this pr contains some small fixes in the test_updater_key_rotations.py module.

**Please verify and check that the pull request fulfills the following
requirements**:

- [x] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [x] Tests have been added for the bug fix or new feature
- [x] Docs have been added for the bug fix or new feature

Commit:
Rename RootVersions to MdVersions

RootVersions can be used for key rotation tests beside root and as such
it makes sense to rename it to something more generic.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

Commit:
Rename expected_result to expected_error

Rename expected_result to expected_error as the value expected is never
a true result in the current test cases, but it's an expected error.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

Commit:
test key-rotations: setup & teardown to class func

Make setup and teardown classmethods. There is no reason why we would
want to create the signers and keys for each test case and this only
slows the test module execution.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

Commit:
Tests: add tests for non-root key rotations

The amount of non-root key rotations is lower than the one for root as
the repository stores all root versions in order to be able to do
rollback check compared to timestamp, snapshot, and targets where it
stores only one version.

All of the root roles are needed in those tests again as root
delegates the keys for each of the other three top-level metadata roles.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

MVrachev commented:
I have rebased on top of develop.
@jku can you have a look?

jku commented on tests/test_updater_key_rotations.py:
```diff
@@ -10,20 +10,20 @@
 
 @dataclass
-class RootVersion:
+class MetadataVersion:
```
The lines are already long, this makes them even longer. Maybe `MdVersion` or just `Version`?

jku commented on tests/test_updater_key_rotations.py:
```diff
@@ -36,23 +36,25 @@ class TestUpdaterKeyRotations(unittest.TestCase):
+    def setUpClass(cls) -> None:
+        cls.sim: RepositorySimulator
+        cls.metadata_dir: str
+        cls.subtest_count = 0
```
subtest_count doesn't look like class variable, It seems more like a test specific variable to me

Now `subtest_count` is admittedly a hack that's used to create directory names when dumping repository state (because the subtest name does not seem to be available during the test) so I will also happily take a solution that removes this variable altogether... 

One option I considered but did not bother testing was to use `str(md_version)`  (where md_version is the subtests input) as the directory name instead of subtest_count  -- that should be a legal directory name on linux at least? If you want to check if that works, I'm fine with trying that

MVrachev replied:
I have another workaround.
See https://github.com/theupdateframework/python-tuf/pull/1691/commits/51f6baf57df988e711e0a3db9e99a8f197c68521

jku replied:
that is very clever

jku commented on tests/test_updater_key_rotations.py:
```diff
@@ -188,20 +192,96 @@ def test_root_rotation(self, root_versions: List[RootVersion]) -> None:
+        roles = ["timestamp", "snapshot", "targets"]
+        for role in roles:
+
+            # clear root keys, signers
```
```suggestion
            # clear role keys, signers
```

jku commented on tests/test_updater_key_rotations.py:
```diff
@@ -188,20 +192,96 @@ def test_root_rotation(self, root_versions: List[RootVersion]) -> None:
+            if expected_error is None:
+                self._run_refresh()
+
+                # Call fetch_metadata to sign metadata with new keys
+                expected_local_md: Metadata = self.sim._fetch_metadata(role)
+                # assert local metadata role is on disk as expected
+                md_path = os.path.join(self.metadata_dir, f"{role}.json")
+                with open(md_path, "rb") as f:
+                    data = f.read()
+                    self.assertEqual(data, expected_local_md)
+                    # md.version = 1 as only 1 version has been published
+                    md = Metadata.from_bytes(data)
+                    self.assertEqual(md.signed.version, 1)
```
I wonder if these checks are useful, or if checking that correct files exist is enough?

MVrachev replied:
I was wondering the same while I was writing.
The part that is in question for me is the version check as I am not sure if there is a case when the version can end up something different than 1.
The other checks seem okay for me - checking that the local metadata file exists and it's with the content we expect.
Will remove the version check and then we can discuss it again.

MVrachev replied:
Removed the version check. What do you think about the rest of the checks?

jku reviewed (commented):
I wonder if the `for role in roles` loop really needs to contain everything it does? Publishing one new root version (that includes the changes for all three metadata) and refreshing once should be enough?

modifying keys, threshold and sigs does need to be in a loop, but nothing else is _really_ required. If the resulting file contents are checked, that needs to happen in a loop as well but I wonder if just checking that the files are there might be enough (assuming subtest_setup removed the files in between)?

Commit:
Tests decorator: save case name

Save the case name in the "unittest.TestCase" object when executing
test cases through the "run_sub_tests_with_dataset" decorator.
This allows dumping directories with names specific to the test case
that can be used for debugging as showed in
test_updater_key_rotations.py

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

MVrachev commented:
@jku I addressed your comments and the pr is ready for another review.

MVrachev commented:
Rebased on top of latest develop changes because of a conflict.

jku reviewed (approved):
The test case name trick for dumps is very clever :clap: (I hope others find the --dump feature, it really makes it easier to see what's going on)

The loop in the test makes it a little difficult to figure out what is actually being tested but it does seem to work correctly. I remember we discussed possibly implementing three separate functions for timestamp/snapshot/targets (that would use the same data and maybe same helper function): that would probably be easier to understand ... but maybe it's not worth all that work: this version works just fine.

LGTM, thanks Martin
